### ISSUE TITLE
Force mount ZFS Root on Boot not working

### ISSUE THREAD

**Issue by UralZima** [zfsonlinux/zfs#5813]
Hello all. I am sorry if this question is related only to gentoo, but I think that processing kernel arguments is ZFS's job.

I use ZFS on production remote gentoo server. All was fine, until my server freezed and I needed to reset it. It failed with 'device or resource busy' error on mounting rootfs. The simply action need to avoid this - import and export it. But for me it was really hard. remote management virtual cd mount is very slow, buggy, and what I did was boot from live cd, import -f pool and export it. But why it is not forced at boot? Why I need to do these additional steps?

my kernel line is: linux /kernel-genkernel-x86_64-4.9.11-hardened crypt_root=/dev/sda2  real_root=ZFS=mypool/ROOT/rootfs dozfs=force by=id net.ifnames=0 zfsforce=1 

zfs version is 0.6.5.9

even when I updated kernel and did a normal reboot, first time it again booted with 'device or resource busy' error.

I have two 'force' keywords, and it ignores this! What to do?

Thanks

**Comment by AndCycle** [comment]
I don't get the problem on my gentoo zfs setup, 

when panic reboot it just fine with `dozfs=force` , 
you can read `dozfs` from `init` use by `initrd` as it just add `-f` arg for `/etc/initrd.scripts` ,

I think you got other problem than simple force import can do,
as it already suggest `device or resource busy` which is not something force import can fix.

**Comment by UralZima** [comment]
The problem is only in force import. 
When I boot livecd, and do usual import, it reports the same. When I import -f and then export, I can do normal input again. 
I use ZFS with Gentoo more than year, and had a few times with such errors while I forced import. And it was usually after hard reset. 

I looked at the initrd scripts and it must work fine, however I didn't found them on my /etc/, maybe they are in /usr/share somewhere. 

Last time I got this error in mounting mypool/ROOT/rootfs, not the whole import. The error was especially it can't mount mypool/ROOT/rootfs on /newroot. And dropped to the initramfs shell

I got this error twice last day, and only import -f/export helped... Any suggestions?

**Comment by AndCycle** [comment]
`/etc/initrd.scripts` is in your `initrd`, not your real_root, 
you need to extract them from your `initrd` to view it's content,

and you should be able to see it when you dropped into ` initramfs shell`,

have you tried to import with -f when you dropped into ` initramfs shell`?

**Comment by AndCycle** [comment]
by the way, there is a related bug could cause import issue
https://bugs.gentoo.org/show_bug.cgi?id=532746

as I set my hostid to all null in order to avoid the endian swap issue.

**Comment by UralZima** [comment]
I unpacked my initramfs and checked these files, they looks like on master.

unfortunately IPMI keyboard is not working in initramfs, however in ubuntu works. I enabled all USB/HID drivers, but it does not work... Maybe some missing firmware. So it just drops me to a prompt, where I need to type 'shell' or path to another rootfs, instead of mypool/ROOT/rootfs. As I said, it is very rare cases when I get such error. but it happens, and on remote server it is very painfully.

But. I just reviewed the source code, and we see, that ZPOOL_IMPORT var is used in initrd.scripts. But is defined in /linuxrc, and AFTER source'ing /etc/initrd.scripts, so the variable is defined later than trying to use it, am I right in that? Maybe it is better to move ZPOOL_IMPORT somewhere to initrd.defaults... I am not sure about this bug on gentoo's bugtracker
